wallet/src/wallet/mod.rs

```diff
@@ -347,8 +365,18 @@ impl<B: storage::Backend> Wallet<B> {
 
         match version {
             WALLET_VERSION_UNINITIALIZED => return Err(WalletError::WalletNotInitialized),
-            WALLET_VERSION_V1 => Self::migration_v2(db, chain_config)?,
-            WALLET_VERSION_V2 => Self::migration_v3(db, chain_config)?,
+            WALLET_VERSION_V1 => {
+                Self::migration_v2(db, chain_config.clone())?;
+                Self::migration_v3(db, chain_config)?;
+                Self::migration_v4(db)?;
+            }
+            WALLET_VERSION_V2 => {
+                Self::migration_v3(db, chain_config)?;
+                Self::migration_v4(db)?;
+            }
+            WALLET_VERSION_V3 => {
+                Self::migration_v4(db)?;
+            }
```

Comment:
A recursive or iterative function that checks then runs again would do this more reliably. Basically you need a function that runs one upgrade, and then you run a loop that keeps upgrading until the last version is reached.